client/src/components/AuthProvider.tsx

```diff
@@ -0,0 +1,148 @@
+import { createContext, ReactNode, useContext, useEffect, useState } from 'react';
+import { Alert, CircularProgress, Stack } from '@mui/material';
+import { INFOBIP_API_BASE_URL, CLIENT_ID, REDIRECT_URL } from '../const';
+
+export interface AuthContext {
+  token: string;
+  token_type: string;
+  expires_in: number;
+  accountKey: string;
+  userKey: string;
+  username: string;
+  email: string;
+  locale: string;
+}
+
+const authContext = createContext<AuthContext | null>(null);
+
+export function useAuthContext(): AuthContext {
+  const context = useContext(authContext);
+  if (!context) {
+    throw new Error('useAuthContext can be used only within AuthProvider');
+  }
+
+  return context;
+}
+
+export function AuthProvider({ children }: { children: ReactNode }) {
+  const [authContextState, setAuthContextState] = useState<AuthContext | null>(null);
+  const [isLoading, setLoading] = useState(false);
+  const [error, setError] = useState<any | null>(null);
+
+  useEffect(() => {
+    const params = new URLSearchParams(window.location.search);
+    const code = params.get('code');
+    const state = params.get('state');
+    const conversationId = params.get('conversationId');
+
+    if (authContextState) {
+      return;
+    }
+
+    if (code) {
+      // 2nd step - after redirect back to the app with access code param in the URL, retrieve token with user info
+      setLoading(true);
+      getToken(code)
+        .then(response => {
+          setAuthContextState(response);
+          removeCodeFromLocation();
+          setError(null);
+          setLoading(false);
+        })
+        .catch(error => {
+          setError(`An error occurred while getting token: ${error}`);
+          setLoading(false);
+        });
+    } else if (state) {
+      // 1st step - redirect to authorization endpoint in order to get access code using state param
+      setLoading(true);
+      try {
+        authorizeWithRedirectBack(state, conversationId);
```

Comment:
`authorizeWithRedirectBack` calls a promise inside and awaits it (line 108). But here it'll return before the promise result arrives. We will never catch an error if it is emitted by that promise. Should we turn `authorizeWithRedirectBack` into async function and handle it like `getToken` above?

Reply:
A good catch. Reworked to be more readable and moved logic with redirect out of the async function. Originally it was done differently (inside React component and not outside) and then I refactored it, but without proper testing. After I did everything, I was reading about promises, because I realized that I don't know how they work, but I should read it before I coded :)